clang/runtime/dpct-rt/include/dpct/sparse_utils.hpp

```diff
@@ -1458,6 +1569,224 @@ inline void csr2csc(sycl::queue queue, int m, int n, int nnz,
+/// have the in_order property when using the USM mode.
+/// \param [in] trans_a The operation applied to the matrix A.
+/// \param [in] trans_b The operation applied to the matrix B.
+/// \param [in] m The rows number of op(A) and C.
+/// \param [in] n The columns number of op(B) and C.
+/// \param [in] k The columns number of op(A) and rows number of op(B).
+/// \param [in] info_a Matrix info of the matrix A.
+/// \param [in] nnz_a Non-zero elements number of matrix A.
+/// \param [in] val_a An array containing the non-zero elements of the matrix A.
+/// \param [in] row_ptr_a An array of length row number + 1.
+/// \param [in] col_ind_a An array containing the column indices in index-based
+/// numbering.
+/// \param [in] info_b Matrix info of the matrix B.
+/// \param [in] nnz_b Non-zero elements number of matrix B.
+/// \param [in] val_b An array containing the non-zero elements of the matrix B.
+/// \param [in] row_ptr_b An array of length row number + 1.
+/// \param [in] col_ind_b An array containing the column indices in index-based
+/// numbering.
+/// \param [in] info_c Matrix info of the matrix C.
+/// \param [in] row_ptr_c An array of length row number + 1.
+/// \param [out] nnz_host_ptr Non-zero elements number of matrix C.
+template <typename T>
+void csrgemm_nnz(descriptor_ptr desc, oneapi::mkl::transpose trans_a,
+                 oneapi::mkl::transpose trans_b, int m, int n, int k,
+                 const std::shared_ptr<matrix_info> info_a, int nnz_a,
+                 const T *val_a, const int *row_ptr_a, const int *col_ind_a,
+                 const std::shared_ptr<matrix_info> info_b, int nnz_b,
+                 const T *val_b, const int *row_ptr_b, const int *col_ind_b,
+                 const std::shared_ptr<matrix_info> info_c, int *row_ptr_c,
+                 int *nnz_host_ptr) {
+  using Ty = typename ::dpct::detail::lib_data_traits_t<T>;
+  sycl::queue &queue = desc->get_queue();
+  detail::csrgemm_args_info args(trans_a, trans_b, m, n, k, info_a, val_a,
+                                 row_ptr_a, col_ind_a, info_b, val_b, row_ptr_b,
+                                 col_ind_b, info_c, row_ptr_c);
+  auto &info = desc->get_csrgemm_info_map()[args];
+
+#ifdef DPCT_USM_LEVEL_NONE
+  auto row_ptr_a_buf = dpct::detail::get_memory<int>(row_ptr_a);
+  auto col_ind_a_buf = dpct::detail::get_memory<int>(col_ind_a);
+  auto val_a_buf = dpct::detail::get_memory<Ty>(val_a);
+  auto row_ptr_b_buf = dpct::detail::get_memory<int>(row_ptr_b);
+  auto col_ind_b_buf = dpct::detail::get_memory<int>(col_ind_b);
+  auto val_b_buf = dpct::detail::get_memory<Ty>(val_b);
+  auto row_ptr_c_buf = dpct::detail::get_memory<int>(row_ptr_c);
+  auto col_ind_c_buf = dpct::detail::get_memory<int>(nullptr);
+  auto val_c_buf = dpct::detail::get_memory<Ty>(nullptr);
+#endif
+
+  if (info.is_empty) {
+    auto matrix_handle_a = std::make_unique<detail::matrix_handle_manager_t>(
```

Comment:
We can construct the handle manager in the map directly.
info.matrix_handle_a = detail::matrix_handle_manager_t>(...)